_source/_posts/2022-03-15-programming-security-and-why-rust.md

```diff
@@ -0,0 +1,157 @@
+
+What is safe programming, or to be more precise, what does being safe mean for a programming language? or rather what does unsafe mean? Let's set the context first.
+
+{% include toc.md %}
+
+# Programming safety
+
+> Programming Safety = Memory safety + Type safety + Thread safety
+
+Safety can be categorized into three, memory safety, type safety, and thread safety (or four if you count null safety as a separate item)
+
+## Memory safety
+
+In a memory-safe language, when you access a variable or an item in an array, you can be sure that you are indeed accessing what you meant to or are allowed to access. In other words, you will not be reading or writing into the memory of another variable or pointer by mistake, regardless of what you do in your program.
+
+So, why is this a big deal? Doesn't all major programming languages ensure this?
+
+Yes, to varying extents. But some languages are unsafe by default—for example, C and C++. In C or C++, you can access the memory of another variable by mistake, or you can free a pointer twice; that's called double-free error. Sometimes a program continues to use a pointer after it has been freed, and that's called free after use errors or a dangling pointer error. Such behavior is categorized as undefined behavior as they are unpredictable and causes security vulnerability rather than just crashing the program. In this scenario, a crashing program is a good thing as it won't cause a security vulnerability.
+
+> I call it my billion-dollar mistake. It was the invention of the null reference in 1965
+>
+> - Tony Hoare
+
+Then there is also null safety which is kind of related to memory safety. I come from a Java/JavaScript background, and we are so used to the concept of null. Infamous for being the worst invention in programming. Garbage collected languages need a concept of nothing so that a pointer can be freed when unused. But it also leads to issues and pain, like the null pointer exceptions. Technically this falls under memory safety, but most memory-safe languages still let you use null as a value leading to null pointer errors.
+
+## Type safety
+
+In a type-safe language, when you access a variable, you access it as the correct type of data it is stored as. This gives us the confidence to work on data without manually checking for the data type during runtime. Memory safety is required for a language to be type-safe.
+
+## Thread safety
+
+In a thread-safe language, you can access or modify the same memory from multiple threads simultaneously without worrying about data races. This is generally achieved using message passing techniques, mutual exclusion locks (Mutex), and thread synchronization. Thread safety is required for optimal memory and type safety, so generally, memory and type-safe languages tend to be thread-safe as well.
```

Comment:
```suggestion

Without thread safety, one thread could write a variable at the same time another is reading it, leading to undefined behavior, such as crashes or security vulnerabilities. When threads are accessing and modifying the same memory simultaneously, we call it a race condition or data race. 

In a thread-safe language, you can access or modify the same memory from multiple threads simultaneously without worrying about data races. This is generally achieved using message passing techniques, mutual exclusion locks (Mutex), and thread synchronization. Thread safety is required for optimal memory and type safety, so generally, memory and type-safe languages tend to be thread-safe as well.
```

Reply:
i'll take this with some modification since crashes are hard to categorize as undefined behavior since there will be panics in that case